docs/devGuide/migratingToTypeScript.md

```diff
@@ -0,0 +1,135 @@
+   Default import | `const x = require(y)`     | `import x = require(y)`                                          | `import x from y`
+   Named import   | `const { x } = require(y)` | N/A, workaround:<br>`import z = require(y)`<br>`const { x } = z` | `import { x } from y`
+   Named export   | `module.exports = { x }`   | `export = { x }`                                                 | `export { x }`
+   Default export | `module.exports = x`       | `export = x`                                                     | `export default x`
+
+   Similar to the process of refactoring a function, you might have to **adjust how these files are imported by other TypeScript (`.ts`) files**. If you are changing it to the TypeScript equivalent syntax, then it must be imported with so as well. The same goes for the ES6 syntax. 
+
+   Start with the TypeScript equivalent syntax, as it is compiled to be exactly the same as the CommonJS syntax (except the named import case), making it the safest choice when adapting the file. You can read more about this syntax in the [official documentation](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require).
+
+   The ES6 syntax compiles into something more than just the CommonJS syntax when compiled, and might lead to unexpected behaviours when another `.js` file imports the compiled file, to the point of failing tests in some cases. You can try changing to the ES6 syntax, but only proceed if there are no failing tests. The TypeScript equivalent syntax is enough for the initial adaptation.
+
+   However, in some cases, changing to the ES6 syntax might be necessary. For instance, you cannot use the TypeScript equivalent syntax if you want to import/export TypeScript specific constructs, such as types, interfaces, and enums. Again, when doing so, make sure that all tests pass.
+
+   <box type="info">
+
+   We will perform a mass change into ES6 syntax after all files in core have been adapted into TypeScript.
+   </box>
+
+3. **Adapt the files** to fully use TypeScript.
+
+   The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Only fix what is necessary and do not modify any code functionality. Avoid using `any` as best as you can.
+
+   If you happen to encounter type errors related to using names from an internal dependency that is still in JavaScript (`.js`), you can try to infer simple types in your file first as a "stand-in" of the more robust types when that internal dependency is eventually migrated.
+
+   On the flip side of the above situation, as you now have developed a robust type for your own file, you might have to **adjust how other TypeScript (`.ts`) files type names referenced from your files**. Some type declarations might need to be made stricter, or can be removed if it's not necessary anymore, and so on. This will ensure that the files will be incrementally stricter. 
```

Comment:
```suggestion
   On the flip side of the above situation, once you have developed a robust type for your own file, **adjust how other TypeScript (`.ts`) files type names referenced from your files** as applicable. Some type declarations might need to be made stricter, or can be removed if it's not necessary anymore, and so on. This will ensure that the files will be incrementally stricter. 
```